interfaces/builtin/mount_control.go

```diff
@@ -0,0 +1,337 @@
+// separate rules; otherwise, one could specify a path like
+//   $SNAP/foo, /** rw
+// and get read-write access to the whole filesystem.
+var whereRegexp = regexp.MustCompile(`^(/media|\$SNAP|\$SNAP_COMMON|\$SNAP_DATA)/[^,]*$`)
+
+// Excluding spaces and other characters which might allow constructing a
+// malicious string like
+//   auto) options=() /malicious/content /var/lib/snapd/hostfs/...,\n mount fstype=(
+var typeRegexp = regexp.MustCompile(`^[a-z0-9-]+$`)
+
+type MountInfo struct {
+	what       string
+	where      string
+	persistent bool
+	types      []string
+	options    []string
+}
+
+func parseStringList(mountEntry map[string]interface{}, fieldName string) ([]string, error) {
+	var list []string
+	value, ok := mountEntry[fieldName]
+	if ok {
+		interfaceList, ok := value.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf(`mount-control "mount.%s" must be an array of strings (got %q)`, fieldName, value)
+		}
+		for _, iface := range interfaceList {
+			valueString, ok := iface.(string)
+			if !ok {
+				return nil, fmt.Errorf(`mount-control "mount.%s" element not a string (%q)`, fieldName, iface)
+			}
+			list = append(list, valueString)
+		}
+	}
+	return list, nil
+}
+
+func enumerateMounts(plug interfaces.Attrer, fn func(mountInfo *MountInfo) error) error {
+	mountAttr, ok := plug.Lookup("mount")
+	if !ok {
+		return nil
+	}
+	mounts, ok := mountAttr.([]interface{})
+	if !ok {
+		return errors.New(mountAttrNotADictionary)
+	}
+
+	for _, m := range mounts {
+		mount, ok := m.(map[string]interface{})
+		if !ok {
+			return errors.New(mountAttrNotADictionary)
+		}
+
+		what, ok := mount["what"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount.what" must be a string`)
+		}
+
+		where, ok := mount["where"].(string)
+		if !ok {
+			return fmt.Errorf(`mount-control "mount.where" must be a string`)
+		}
+
+		persistent := false
+		persistentValue, ok := mount["persistent"]
+		if ok {
+			if persistent, ok = persistentValue.(bool); !ok {
+				return fmt.Errorf(`mount-control "mount.persistent" must be a boolean`)
+			}
+		}
+
+		types, err := parseStringList(mount, "type")
+		if err != nil {
+			return err
+		}
+
+		options, err := parseStringList(mount, "options")
+		if err != nil {
+			return err
+		}
+
+		mountInfo := &MountInfo{
+			what:       what,
+			where:      where,
+			persistent: persistent,
+			types:      types,
+			options:    options,
+		}
+
+		if err := fn(mountInfo); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func validateWhatAttr(what string) error {
+	if what[0] != '/' {
+		return fmt.Errorf(`mount-control "what" attribute must be an absolute path`)
+	}
+
+	if !cleanSubPath(what) {
+		return fmt.Errorf(`mount-control "what" pattern is not clean: %q`, what)
+	}
+
+	if _, err := utils.NewPathPattern(what); err != nil {
+		return fmt.Errorf(`mount-control "what" attribute error: %v`, err)
+	}
+
+	return nil
+}
+
+func validateWhereAttr(where string) error {
```

Comment:
I wonder if this function should also be a little bit more conservative and disallow "where" being specified as some important files in the snap such as ./meta/snap.yaml or ./meta/gadget.yaml since snapd will parse those files. I am only wondering this, since currently in the per-snap mount namespace at least, the mountpoint for $SNAP is a "slave" to the mountpoint for $SNAP in the root mount namespace and so any mounts performed on top of $SNAP will NOT get propagated to the root mount namespace for snapd to get confused, but I am still thinking it might pay for us to be slightly paranoid about this anyways.

Reply:
I removed support for `$SNAP` in the "where" attribute, so this should not be an issue anymore.